horovod/common/ops/collective_operations.cc

```diff
@@ -0,0 +1,401 @@
+// Copyright 2016 The TensorFlow Authors. All Rights Reserved.
+// Modifications copyright (C) 2019 Uber Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+
+#include "collective_operations.h"
+
+namespace horovod {
+namespace common {
+
+HorovodOp::HorovodOp(HorovodGlobalState* global_state) : global_state_(global_state) {}
+
+// Allreduce
+AllreduceOp::AllreduceOp(HorovodGlobalState* global_state) : HorovodOp(global_state) {}
+
+Status AllreduceOp::Execute(std::vector<TensorTableEntry>& entries, const MPIResponse& response) {
+  auto& first_entry = entries[0];
+
+  Initialize(entries, response);
+
+  const void* fused_input_data;
+  void* buffer_data;
+  size_t buffer_len;
+
+  // Copy memory into the fusion buffer.
+  if (entries.size() > 1) {
+    // Access the fusion buffer.
+    auto& buffer = global_state_->fusion_buffer.GetBuffer(
+        first_entry.device, first_entry.context->framework());
+    buffer_data = const_cast<void*>(buffer->AccessData(first_entry.context));
+
+    RecordEventStart(MEMCPY_IN_FUSION_BUFFER, entries);
+    int64_t offset = 0;
+    for (auto& e : entries) {
+      void *buffer_data_at_offset = (uint8_t *) buffer_data + offset;
+      MemcpyInFusionBuffer(buffer_data_at_offset, e, entries);
+      offset += e.tensor->size();
+    }
+
+    StreamSynchronize(entries);
+    RecordEventEnd(MEMCPY_IN_FUSION_BUFFER, entries);
```

Comment:
I got vastly confused here - "why would we synchronize and then record an event?", later to realize that in some cases one of these methods is empty, and in others the other method is empty.  I believe it will be very hard to follow for folks who want to extend Horovod.  I think your original assessment that there is sufficient discrepancy between these methods is correct.

Can we try to refactor operations in a different way - by having a set of helper classes/functions, but allowing each strategy to fully implement the logic that it needs?  Ideally, we'd have just a single level of inheritance (from interface), hopefully at a cost of minimal code duplication.

Reply:
I made a change to replace all this with `StartMemcpyInFusionBuffer` and `EndMemcpyInFusionBuffer`.  Let me know if that makes things more clear.

I'm thinking about how we could best implement the entire logic at the op level without duplicating a bunch of code.  The biggest issue is that all the fusion buffer stuff is completely generic, except for the actual implementation of the memcpy.  It's possible to expose functions that do fusion buffering in and out and passing in function pointers to do the work, but it could get equally messy.

Perhaps we should chat more in person.